presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSortColumns.java

```diff
@@ -0,0 +1,82 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.LogicalProperties;
+import com.facebook.presto.spi.plan.Ordering;
+import com.facebook.presto.spi.plan.OrderingScheme;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.SortNode;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.sort;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Removes sort columns from input if the source has a Key that refers to the ordering columns
+ */
+public class RemoveRedundantSortColumns
+        implements Rule<SortNode>
+{
+    private static final Pattern<SortNode> PATTERN = sort().matching(p -> ((GroupReference) p.getSource()).getLogicalProperties().isPresent());
+
+    @Override
+    public Pattern<SortNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SortNode node, Captures captures, Context context)
+    {
+        OrderingScheme orderingScheme = node.getOrderingScheme();
+
+        LogicalProperties sourceLogicalProperties = ((GroupReference) node.getSource()).getLogicalProperties().get();
+        OrderingScheme newOrderingScheme = pruneOrderingColumns(orderingScheme, sourceLogicalProperties);
+
+        if (newOrderingScheme.equals(orderingScheme)) {
+            return Result.empty();
+        }
+
+        return Result.ofPlanNode(new SortNode(node.getSourceLocation(), node.getId(), node.getStatsEquivalentPlanNode(), node.getSource(), newOrderingScheme, node.isPartial()));
+    }
+
+    static OrderingScheme pruneOrderingColumns(OrderingScheme nodeOrderingScheme, LogicalProperties sourceLogicalProperties)
+    {
+        List<VariableReferenceExpression> orderingVariables = nodeOrderingScheme.getOrderBy().stream().map(Ordering::getVariable).collect(toImmutableList());
+        int sizeSmallestDistinctPrefix = sizeSmallestDistinctPrefix(sourceLogicalProperties, orderingVariables);
+        if (sizeSmallestDistinctPrefix == 0) {
+            return nodeOrderingScheme;
+        }
+
+        List<Ordering> keyPrefix = nodeOrderingScheme.getOrderBy().subList(0, sizeSmallestDistinctPrefix);
+        return new OrderingScheme(keyPrefix);
+    }
+
+    private static int sizeSmallestDistinctPrefix(LogicalProperties logicalProperties, List<VariableReferenceExpression> candidateVariables)
+    {
+        for (int prefixSize = 1; prefixSize <= candidateVariables.size(); prefixSize++) {
+            if (logicalProperties.isDistinct(ImmutableSet.copyOf(candidateVariables.subList(0, prefixSize)))) {
```

Comment:
Doesn't need a copy right? In fact we probably should change LogicalProperties::isDistinct to take in an ImmutableSet

Reply:
`isDistinct` needs an input Set -
https://github.com/prestodb/presto/blob/5b1f2ddef7d03568941fab44daeb9c4cc7cfaaf9/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java#L148-L152

Note sublist won't make a copy itself, that is just a view
I'll make a single HashSet and add to it instead, avoid the cost of making a new set everytime